### PR TITLE
Cumulus 3187

### PR DESCRIPTION
**Pull request by Nnaga1**
**Summary:** Summary of changes

Fixed TS issues in 3097 code while also removing the EarthdataLogin class to instead export individual methods

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [x ] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 1]
```diff
@@ -1,202 +1,155 @@
...
-   * @returns {Promise.<string>} the token
-   */
-  async getEDLToken(): Promise<string> {
-    let token = await this.retrieveEDLToken();
-    if (token === undefined) {
-      token = await this.createEDLToken();
-    }
-    return token;
+/**
+* The method for getting the Earthdata Login endpoint URL based on the EDL environment
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:30]: The indentation of these JSDocs needs to be fixed. The `*` here should be under the first `*` above.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 2]
```diff
@@ -1,202 +1,155 @@
...
-   *
-   * @throws {Error} - EarthdataLogin error
-   */
-  handleHttpError(error: Error) {
-    const statusCode = get(error, 'response.statusCode', error.code);
-    const statusMessage = get(error, 'response.statusMessage', error.message);
-    const responseErrorDescription = JSON.parse(get(error, 'response.body')).error_description;
-    const errorMessage = `EarthdataLogin error: ${responseErrorDescription},  statusCode: ${statusCode}, statusMessage: ${statusMessage}. Earthdata Login Request failed`;
+/**
+* This helper method is called in the functions that retrieve, create, and revoke EarthdataLogin
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:48]: The indentation of these JSDocs needs to be fixed. The `*` here should be under the first `*` above.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 3]
```diff
@@ -1,202 +1,155 @@
...
+  const statusCode = error.response.statusCode;
+  const statusMessage = error.response.statusMessage || 'Unknown';
+  const errorBody = requestType === 'retrieve' || requestType === 'create' ? JSON.stringify(error.response.body) : error.response.body;
+  const message = `EarthdataLogin error: ${errorBody},  statusCode: ${statusCode}, statusMessage: ${statusMessage}. Earthdata Login Request failed`;
+  return new Error(message);
+};
 
-    throw new Error(errorMessage);
+/**
+  * The method for getting the token from the Earthdata Login endpoint. Sends a GET request
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:63]: The indentation of these JSDocs needs to be fixed. The `*` here should be under the first `*` above.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 4]
```diff
@@ -1,202 +1,155 @@
...
-
-    for (let i = 0; i < Object.keys(response).length; i += 1) {
-      const responseDate = new Date(response[i].expiration_date);
-      if (currDate < responseDate) {
-        return response[i].access_token;
-      }
-    }
-    return undefined;
+/**
+  * The method for creating Earthdata Login token. This method sends a POST request
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:93]: The indentation of these JSDocs needs to be fixed. The `*` here should be under the first `*` above.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 5]
```diff
@@ -1,202 +1,155 @@
...
-          headers: {
-            Authorization: `Basic ${buff}`,
-          },
-        }).json();
-    } catch (error) {
-      this.handleHttpError(error);
-    }
-    return response[0].access_token;
+/**
+  * This method is used for the cmrTokenSpec integration test in order to revoke the
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:120]: The indentation of these JSDocs needs to be fixed. The `*` here should be under the first `*` above.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 6]
```diff
@@ -1,202 +1,155 @@
...
-    } catch (error) {
-      this.handleHttpError(error);
-    }
+/**
+ * The method that determines if a user has a token, to call the retrieveEDLToken function
+ * which makes an API call to the Earthdata Login endpoint, or to create the token with the
+ * createEDLToken function. Returns the token as a string.
+ *
+ * @returns {string} the token
+*/
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:147]: The indentation of these JSDocs needs to be fixed. The `*` here should be under the `*` above.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 7]
```diff
@@ -1,202 +1,155 @@
...
-      default:
-        return 'https://sit.urs.earthdata.nasa.gov';
-    }
-  }
+/** This method is for parsing a caught error which is not an HTTPerror
+ * in case the EDL endpoint call results in an unexpected error
+ *
+ * @returns {Error}
+ */
+const parseCaughtError = (e: unknown): Error => (e instanceof Error ? e : new Error(`${e}`));
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:27]: Let's move this to `packages/common/src/errors.ts` so that it can be used elsewhere.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 8]
```diff
@@ -1,202 +1,155 @@
...
-  async getEDLToken(): Promise<string> {
-    let token = await this.retrieveEDLToken();
-    if (token === undefined) {
-      token = await this.createEDLToken();
-    }
-    return token;
+/**
+* The method for getting the Earthdata Login endpoint URL based on the EDL environment
+*
+* @returns {<string>} the endpoint URL
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:32]:
```suggestion
* @returns {string} the endpoint URL
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 9]
```diff
@@ -1,202 +1,155 @@
...
+*/
+const getEdlUrl = (env: string) : string => {
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:34]:
```suggestion
const getEdlUrl = (env: string): string => {
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 10]
```diff
@@ -1,202 +1,155 @@
...
 
-    throw new Error(errorMessage);
+/**
+  * The method for getting the token from the Earthdata Login endpoint. Sends a GET request
+  * with the users' base64 encoded username and password as a header for authorization. If the
+  * users' credentials are accepted the first unexpired token is retrieved, if one exists, and
+  * returned on their behalf, if not, an error is thrown. If the user does not have a token
+  * in Earthdata Login then undefined is returned to indicate to token needs to be created.
+  *
+  * @returns {string} the token
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:69]: This actually returns `Promise<string|undefined>`, not `string`. Let's just remove the `@returns` tag. The TS code itself documents the correct type.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 11]
```diff
@@ -1,202 +1,155 @@
...
+*/
+export const retrieveEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string | undefined> => {
+  // response: get a token from the Earthdata login endpoint using credentials if exists
+  let rawResponse: any;
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:74]: Using `any` disables type checking. The correct type in this case is `Response<unknown>`. You'll need to add `Response` to the `got` import at the top of the file.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 12]
```diff
@@ -1,202 +1,155 @@
...
+  try {
+    rawResponse = await got.get(`${getEdlUrl(edlEnv)}/api/users/tokens`,
+      {
+        responseType: 'json',
+        username: username,
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:79]:
```suggestion
        username,
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 13]
```diff
@@ -1,202 +1,155 @@
...
+        password: password,
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:80]:
```suggestion
        password,
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 14]
```diff
@@ -1,202 +1,155 @@
...
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'retrieve');
+    throw parseCaughtError(error);
   }
+  const tokens = GetTokenResponseBody.parse(rawResponse.body);
+  const isTokenExpired = (token: Token) => new Date(token.expiration_date) < new Date();
+  const unExpiredTokens = tokens.filter((token: any) => !isTokenExpired(token));
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:88]:
```suggestion
  const unExpiredTokens = tokens.filter((token: Token) => !isTokenExpired(token));
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 15]
```diff
@@ -1,202 +1,155 @@
...
+  * The method for getting the token from the Earthdata Login endpoint. Sends a GET request
+  * with the users' base64 encoded username and password as a header for authorization. If the
+  * users' credentials are accepted the first unexpired token is retrieved, if one exists, and
+  * returned on their behalf, if not, an error is thrown. If the user does not have a token
+  * in Earthdata Login then undefined is returned to indicate to token needs to be created.
+  *
+  * @returns {string} the token
+*/
+export const retrieveEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string | undefined> => {
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:72]: It's a little easier to read if it's formatted like this:

```suggestion
export const retrieveEDLToken = async (
  username: string,
  password: string,
  edlEnv: string
): Promise<string | undefined> => {
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 16]
```diff
@@ -1,202 +1,155 @@
...
+        password: password,
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'retrieve');
+    throw parseCaughtError(error);
   }
+  const tokens = GetTokenResponseBody.parse(rawResponse.body);
+  const isTokenExpired = (token: Token) => new Date(token.expiration_date) < new Date();
+  const unExpiredTokens = tokens.filter((token: any) => !isTokenExpired(token));
+  return unExpiredTokens.length > 0 ? sortBy(unExpiredTokens, ['expiration_date'])[0].access_token : undefined;
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:89]: Assuming the test data is correct, and you are looking for the token that expires last, then this line is not going to give you the behavior you're expecting. For example:

```
>   sortBy(
...     [
...       { expiration_date: '1/1/2000' },
...       { expiration_date: '2/1/2001' },
...     ],
...     'expiration_date'
...   )
[ { expiration_date: '1/1/2000' }, { expiration_date: '2/1/2001' } ]
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 17]
```diff
@@ -1,202 +1,155 @@
...
-    }
-    return undefined;
+/**
+  * The method for creating Earthdata Login token. This method sends a POST request
+  * to the Earthdata Login endpoint URL in order to create a token for the user. The users'
+  * username and password are sent as base64 encoded credentials as a header for
+  * authorization. If the users' credentials are accepted a token is created on their
+  * behalf and returned, if not, an error is thrown.
+  *
+  * @returns {string} the token
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:99]:
```suggestion
  * @returns {Promise<string>} the token
```

Like I mentioned in a similar comment, it would be better to just remove this `@returns` tag rather than risk an incorrect value.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 18]
```diff
@@ -1,202 +1,155 @@
...
+*/
+export const createEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string> => {
+  let rawResponse: any;
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:103]:
```suggestion
  let rawResponse: Response<unknown>;
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 19]
```diff
@@ -1,202 +1,155 @@
...
+  * The method for creating Earthdata Login token. This method sends a POST request
+  * to the Earthdata Login endpoint URL in order to create a token for the user. The users'
+  * username and password are sent as base64 encoded credentials as a header for
+  * authorization. If the users' credentials are accepted a token is created on their
+  * behalf and returned, if not, an error is thrown.
+  *
+  * @returns {string} the token
+*/
+export const createEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string> => {
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:102]: Easier to read as:

```suggestion
export const createEDLToken = async (
  username: string,
  password: string,
  edlEnv: string
): Promise<string> => {
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 20]
```diff
@@ -1,202 +1,155 @@
...
-  *
-  * @example
-  *
-  * {
-  *  "username": "janedoe",
-  *  "password": "password",
-  *  "edlEnv": "UAT"
-  * }
-  */
+const PostTokenResponseBody = z.array(PostTokenSchema);
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:20]: How many elements will this array have? 0, 1, or more? If the answer is "always 1", then we should define this as `z.tuple([PostTokenSchema])`.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 21]
```diff
@@ -1,202 +1,155 @@
...
+        responseType: 'json',
+        username: username,
+        password: password,
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'create');
+    throw parseCaughtError(error);
   }
+  const response = PostTokenResponseBody.parse(rawResponse.body);
+  return response[0].access_token;
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:116]: Without a check for the number of elements in `response`, `response[0]` is potentially `undefined`.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 22]
```diff
@@ -1,202 +1,155 @@
...
+  * @returns {string} the token
+*/
+export const createEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string> => {
+  let rawResponse: any;
+  try {
+    rawResponse = await got.post(`${getEdlUrl(edlEnv)}/api/users/token`,
+      {
+        responseType: 'json',
+        username: username,
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:108]:
```suggestion
        username,
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 23]
```diff
@@ -1,202 +1,155 @@
...
+        password: password,
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:109]:
```suggestion
        password,
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 24]
```diff
@@ -1,202 +1,155 @@
...
-      this.handleHttpError(error);
-    }
-    return response[0].access_token;
+/**
+  * This method is used for the cmrTokenSpec integration test in order to revoke the
+  * token that is created for testing.
+  *
+*/
+export const revokeEDLToken = async (username: string, password: string, edlEnv: string,
+  token: string): Promise<void> => {
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:125]:
```suggestion
export const revokeEDLToken = async (
  username: string,
  password: string,
  edlEnv: string,
  token: string
): Promise<void> => {
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 25]
```diff
@@ -1,202 +1,155 @@
...
+  try {
+    await got.post(`${getEdlUrl(edlEnv)}/api/users/revoke_token`,
+      {
+        searchParams: {
+          token,
+        },
+        username: username,
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:132]:
```suggestion
        username,
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 26]
```diff
@@ -1,202 +1,155 @@
...
+        password: password,
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:133]:
```suggestion
        password,
```

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 27]
```diff
@@ -1,202 +1,155 @@
...
-      /* eslint-enable @typescript-eslint/no-unused-vars */
-    } catch (error) {
-      this.handleHttpError(error);
-    }
+/**
+ * The method that determines if a user has a token, to call the retrieveEDLToken function
+ * which makes an API call to the Earthdata Login endpoint, or to create the token with the
+ * createEDLToken function. Returns the token as a string.
+ *
+ * @returns {string} the token
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:146]: Same comment as elsewhere, this is a `Promise<string>`, not a `string`.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 28]
```diff
@@ -1,202 +1,155 @@
...
+*/
+export const getEDLToken = async (username: string, password: string, edlEnv: string):
```
Comment (marchuffnagle) [inline packages/cmr-client/src/EarthdataLogin.ts:148]: Same formatting comment as elsewhere.

File: packages/cmr-client/tests/test-EarthdataLogin.js [hunk 29]
```diff
@@ -3,7 +3,13 @@
 const test = require('ava');
```
Comment (marchuffnagle) [inline packages/cmr-client/tests/test-EarthdataLogin.js:3]: Please add a `retrieveEDLToken` test showing that it selects the latest token when the tokens are returned and they are not sorted by `expiration_date`.

File: packages/cmr-client/src/CMR.ts [hunk 30]
```diff
@@ -31,12 +30,10 @@ async function updateToken(
...
 ): Promise<string> {
-  const earthdataLoginObject = new EarthdataLogin({
-    username: username,
-    password: password,
-    edlEnv: process.env.CMR_ENVIRONMENT,
-  });
-  return await earthdataLoginObject.getEDLToken();
+  // eslint-disable-next-line @typescript-eslint/dot-notation
+  const edlEnv = process.env['CMR_ENVIRONMENT'];
+  if (!edlEnv) throw new Error('CMR_ENVIRONMENT not set');
```
Comment (marchuffnagle) [inline packages/cmr-client/src/CMR.ts:35]: You can use the `getRequiredEnvVar` function from `packages/common/src/env.ts` here.

**Review by marchuffnagle — CHANGES_REQUESTED** [review]

File: packages/cmr-client/src/EarthdataLogin.ts (hunk 10 above)
Comment (marchuffnagle) [reply packages/cmr-client/src/EarthdataLogin.ts:69]: The EDL ones were the only ones I noticed that were incorrect. Were there others that you changed?

File: packages/cmr-client/src/EarthdataLogin.ts (hunk 7 above)
Comment (marchuffnagle) [reply packages/cmr-client/src/EarthdataLogin.ts:27]: > I can move the parseHttpError there as well 👍

That one seems pretty specific to Earthdata Login, it should probably be kept here.

File: packages/cmr-client/src/CMR.ts (hunk 30 above)
Comment (marchuffnagle) [reply packages/cmr-client/src/CMR.ts:35]: `getRequiredEnvVar('CMR_ENVIRONMENT')`

File: packages/cmr-client/src/EarthdataLogin.ts (hunk 21 above)
Comment (marchuffnagle) [reply packages/cmr-client/src/EarthdataLogin.ts:116]: If that's the case, then `PostTokenResponseBody` is never going to match. That schema is looking for an array, but the docs you link to here show an object being returned.

It also looks like the tests are using the incorrect value. They are sending an array, but the linked docs show an object being returned.

File: packages/cmr-client/tests/test-EarthdataLogin.js (hunk 29 above)
Comment (marchuffnagle) [reply packages/cmr-client/tests/test-EarthdataLogin.js:3]: Correct. To make sure you're really testing it, make sure that the list you get back is not already sorted. So, for example, return three tokens and make the latest one the one in the middle of the list.

File: packages/cmr-client/src/EarthdataLogin.ts (hunk 16 above)
Comment (marchuffnagle) [reply packages/cmr-client/src/EarthdataLogin.ts:89]: No. The issue is that you're sorting the expiration dates as strings. With the test data that you're using, though, that is not going to sort them in date order.

File: packages/cmr-client/src/CMR.ts [hunk 31]
```diff
@@ -31,12 +30,10 @@ async function updateToken(
   username: string,
   password: string
 ): Promise<string> {
-  const earthdataLoginObject = new EarthdataLogin({
-    username: username,
-    password: password,
-    edlEnv: process.env.CMR_ENVIRONMENT,
-  });
-  return await earthdataLoginObject.getEDLToken();
+  // eslint-disable-next-line @typescript-eslint/dot-notation
```
[the remark this answers is not in the document]
Comment (marchuffnagle) [reply packages/cmr-client/src/CMR.ts:33]: eslint is not reporting an error here. The comment that is disabling the `@typescript-eslint/dot-notation` rule can be removed.

File: packages/cmr-client/src/EarthdataLogin.ts (hunk 21 above)
Comment (marchuffnagle) [reply packages/cmr-client/src/EarthdataLogin.ts:116]: If it's a string then it's likely JSON that needs to be parsed using `JSON.parse`. That being said, the `got` client should be parsing that JSON response if you've specified `responseType: 'json'`.

**Review by jennyhliu — CHANGES_REQUESTED** [review]
Screen Shot 2023-03-02 at 10 18 07 AM

File: example/spec/parallel/testAPI/cmrTokenSpec.js [hunk 32]
```diff
@@ -29,28 +29,23 @@ describe('When using Earthdata Login Token from CMR', () => {
         username: username,
```
Comment (jennyhliu) [inline example/spec/parallel/testAPI/cmrTokenSpec.js:29]:
```suggestion
        username,
```

File: example/spec/parallel/testAPI/cmrTokenSpec.js [hunk 33]
```diff
@@ -29,28 +29,23 @@ describe('When using Earthdata Login Token from CMR', () => {
...
         password: password,
```
Comment (jennyhliu) [inline example/spec/parallel/testAPI/cmrTokenSpec.js:30]:
```suggestion
        password,
```

File: packages/common/src/errors.ts [hunk 34]
```diff
@@ -0,0 +1,9 @@
+/** This method is for parsing a caught error which is not an HTTPerror
+ * in case the EDL endpoint call results in an unexpected error
+ *
+ * @returns {Error}
+ */
+function parseCaughtError(e: unknown): Error {
+  return (e instanceof Error ? e : new Error(`${e}`));
+}
+export = parseCaughtError;
```
Comment (jennyhliu) [inline packages/common/src/errors.ts:9]: Screen Shot 2023-03-02 at 10 47 27 AM

File: CHANGELOG.md [hunk 35]
```diff
@@ -6,6 +6,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
 
+- **CUMULUS-3187**
+  - Restructured Earthdata Login class to be individual methods as opposed to a Class Object
+  - removed typescript checks and reformatted EarthdataLogin code to be more type friendly
```
Comment (jennyhliu) [inline CHANGELOG.md:11]: `removed typescript checks`?  It actually enabled the checks.

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 36]
```diff
@@ -1,202 +1,170 @@
...
-        return 'https://urs.earthdata.nasa.gov';
-      case 'UAT':
-        return 'https://uat.urs.earthdata.nasa.gov';
-      case 'SIT':
-      default:
-        return 'https://sit.urs.earthdata.nasa.gov';
-    }
-  }
+const returnJWTexp = (token: string) : number =>
+  JSON.parse(Buffer.from(JSON.stringify(token).split('.')[1], 'base64').toString()).exp;
```
Comment (jennyhliu) [inline packages/cmr-client/src/EarthdataLogin.ts:29]: Please add doc to help understand what it does, and add example.

Comment (jennyhliu) [reply packages/cmr-client/src/EarthdataLogin.ts:29]: Where is this logic coming from?

File: packages/cmr-client/src/EarthdataLogin.ts [hunk 37]
```diff
@@ -1,202 +1,170 @@
...
+        responseType: 'json',
+        username,
+        password,
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'retrieve');
+    throw parseCaughtError(error);
   }
+  const tokens = GetTokenResponseBody.parse(rawResponse.body);
+  const isTokenExpired = (token: Token) => (returnJWTexp(token.access_token) < +new Date() / 1000);
```
Comment (jennyhliu) [inline packages/cmr-client/src/EarthdataLogin.ts:91]: move this method outside as a standalone method, and add doc.
what's this `+` in `+new Date() / 1000`

File: packages/cmr-client/package.json [hunk 38]
```diff
@@ -35,11 +38,16 @@
...
     "@cumulus/logger": "14.0.0",
     "got": "^11.8.5",
     "lodash": "^4.17.21",
     "public-ip": "^3.0.0",
-    "xml2js": "^0.4.19"
+    "xml2js": "^0.4.19",
+    "zod": "^3.20.2"
+  },
+  "devDependencies": {
+    "jose": "^4.12.0"
```
Comment (jennyhliu) [inline packages/cmr-client/package.json:51]: Is there any reason to use `jose` instead of the existing package `jsonwebtoken` we have?

File: packages/cmr-client/src/EarthdataLogin.ts (hunk 37 above)
Comment (jennyhliu) [reply packages/cmr-client/src/EarthdataLogin.ts]: You can use [Date.now()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now)

File: packages/cmr-client/tests/EarthdataLogin/test-EarthdataLoginToken.js [hunk 39]
```diff
@@ -0,0 +1,345 @@
+'use strict';
```
Comment (jennyhliu) [inline packages/cmr-client/tests/EarthdataLogin/test-EarthdataLoginToken.js:1]: There is only one test for EarthdataLogin, and I feel that there is no need to have the EarthdataLogin directory unless there are multiple tests

File: packages/cmr-client/package.json [hunk 40]
```diff
@@ -34,12 +37,15 @@
   "author": "Cumulus Authors",
   "license": "Apache-2.0",
   "dependencies": {
-    "@cumulus/aws-client": "14.1.0",
-    "@cumulus/errors": "14.1.0",
-    "@cumulus/logger": "14.1.0",
+    "@cumulus/aws-client": "14.0.0",
```
Comment (jennyhliu) [inline packages/cmr-client/package.json:40]: Update packages to 14.1.0.

File: packages/cmr-client/tests/EarthdataLogin/test-EarthdataLoginToken.js (hunk 39 above)
Comment (jennyhliu) [reply packages/cmr-client/tests/EarthdataLogin/test-EarthdataLoginToken.js]: The helper functions go to the test file test-EarthdataLogin.js

File: packages/cmr-client/tests/test-EarthdataLogin.js [hunk 41]
```diff
@@ -18,160 +27,306 @@ test.after.always(() => {
...
 
+  nock('https://sit.urs.earthdata.nasa.gov')
+    .get('/api/users/tokens')
+    .reply(200, buildGetTokensResponse([token]));
+  const result = await retrieveEDLToken(username, password, 'sit');
+
+  t.is(result, token);
+});
+
+test('retrieveEDLToken returns the last-expiring token if there are multiple tokens', async (t) => {
```
Comment (jennyhliu) [inline packages/cmr-client/tests/test-EarthdataLogin.js:231]: suggest to rewrite this test, as Marc said, 3 tokens with the last-expiring in the middle, and retrieveEDLToken should return that one.
You can create the tokens with map function to avoid duplicate code.
Avoid naming like july/june, it makes the test hard to follow.

File: packages/cmr-client/tests/EarthdataLoginUtils.js [hunk 42]
```diff
@@ -0,0 +1,73 @@
...
+  );
+};
+
+const buildCreateTokenResponse = (token) => {
+  const decodedToken = jwt.verify(token, 'test');
+  if (decodedToken.exp === undefined) throw new Error('token exp is undefined');
+
+  const expiration = new Date(decodedToken.exp * 1000);
+
+  const expirationDate = expiration.toLocaleDateString('en', {
```
Comment (jennyhliu) [inline packages/cmr-client/tests/EarthdataLoginUtils.js:35]: Use UTC time when possible.

File: packages/cmr-client/tests/EarthdataLoginUtils.js [hunk 43]
```diff
@@ -0,0 +1,73 @@
...
+    payload,
+    'test',
+    {
+      issuer: 'Earthdata Login',
+    }
+  );
+};
+
+const buildCreateTokenResponse = (token) => {
+  const decodedToken = jwt.verify(token, 'test');
```
Comment (jennyhliu) [inline packages/cmr-client/tests/EarthdataLoginUtils.js:30]: secrete is always 'test'?

File: packages/cmr-client/tests/EarthdataLoginUtils.js [hunk 44]
```diff
@@ -0,0 +1,73 @@
...
+  const payload = {
+    type: 'User',
+    uid: randomId('uid-'),
+    exp: expirationTime,
+    ...params.payload,
+  };
+
+  return jwt.sign(
+    payload,
+    'test',
```
Comment (jennyhliu) [inline packages/cmr-client/tests/EarthdataLoginUtils.js:22]: secrete is always 'test', use randomId('test') instead.

**Review by jennyhliu — APPROVED** [review]
Nice work!
